File: forms/agile-checklist/back-end-with-loco/src/engine/composite_grader.rs
use std::collections::BTreeMap;

use super::flagged_issues::detect_additional_flags;
use super::items::{ItemDef, PRACTICES_ITEMS, STAKEHOLDERS_ITEMS, Section, TEAMS_ITEMS};
use super::maturity_rules::apply_maturity_rules;
use super::types::{AssessmentData, GradingResult, SectionScore};
use super::utils::{band_for, derive_maturity, round2};

const MIN_ANSWERED_FOR_REPORT: u32 = 30;

fn score_section(
    section: Section,
    items: &[ItemDef],
    answers: &BTreeMap<String, String>,
) -> SectionScore {
    let mut yes_count = 0u32;
    let mut no_count = 0u32;
    let mut not_applicable_count = 0u32;
    let mut unanswered_count = 0u32;
    for item in items {
        match answers.get(item.id).map(String::as_str).unwrap_or("") {
            "yes" => yes_count += 1,
            "no" => no_count += 1,
            "not-applicable" => not_applicable_count += 1,
            _ => unanswered_count += 1,
        }
    }
    let applicable_count = yes_count + no_count;
    let percent = if applicable_count == 0 {
        None
    } else {
        Some(round2((yes_count as f64 / applicable_count as f64) * 100.0))
    };
    let band = band_for(percent);
    SectionScore {
        section: section.slug().to_string(),
        yes_count,
        no_count,
        not_applicable_count,
        unanswered_count,
        applicable_count,
        percent,
        band: band.slug().to_string(),
    }
}

/// Pure function: grades an agile-checklist submission.
///
/// Returns a [`GradingResult`] containing:
/// - per-section scores (yes / no / N/A counts, percentage, band)
/// - composite overall percentage (unweighted mean of section percentages)
/// - maturity level (Ad-hoc / Initial / Developing / Mature / Optimising
///   / Insufficient-Data)
/// - fired maturity rules (one per section with coaching narrative)
/// - additional operational flags (autonomy risk, finished-work risk,
///   psychological-safety risk, etc.)
pub fn grade(data: &AssessmentData) -> GradingResult {
    let answers = &data.answers;
    let teams = score_section(Section::Teams, TEAMS_ITEMS, answers);
    let stakeholders = score_section(Section::Stakeholders, STAKEHOLDERS_ITEMS, answers);
    let practices = score_section(Section::Practices, PRACTICES_ITEMS, answers);

    let answered_count = teams.yes_count
        + teams.no_count
        + teams.not_applicable_count
        + stakeholders.yes_count
        + stakeholders.no_count
        + stakeholders.not_applicable_count
        + practices.yes_count
        + practices.no_count
        + practices.not_applicable_count;

    let defined: Vec<f64> = [teams.percent, stakeholders.percent, practices.percent]
        .iter()
        .filter_map(|p| *p)
        .collect();
    let overall_percent = if defined.len() == 3 && answered_count >= MIN_ANSWERED_FOR_REPORT {
        Some(round2(defined.iter().sum::<f64>() / defined.len() as f64))
    } else {
        None
    };
    let maturity = derive_maturity(overall_percent);

    let fired_rules = apply_maturity_rules(&teams, &stakeholders, &practices);
    let additional_flags =
        detect_additional_flags(answers, &teams, &stakeholders, &practices, answered_count);

    GradingResult {
        answered_count,
        teams,
        stakeholders,
        practices,
        overall_percent,
        maturity: maturity.slug().to_string(),
        maturity_label: maturity.label().to_string(),
        fired_rules,
        additional_flags,
        timestamp: chrono::Utc::now().to_rfc3339(),
    }
}

File: forms/agile-checklist/front-end-with-svelte/src/lib/engine/composite-grader.ts
import type {
  AgileChecklist,
  Band,
  GradingResult,
  Maturity,
  SectionScore,
} from './types.js';
import {
  PRACTICES_ITEMS,
  STAKEHOLDERS_ITEMS,
  TEAMS_ITEMS,
  type ItemDef,
  type SectionId,
} from '$lib/config/items.js';
import { applyMaturityRules } from './maturity-rules.js';
import { detectAdditionalFlags } from './flagged-issues.js';

const MIN_ANSWERED_FOR_REPORT = 30;

export function bandFor(percent: number | null): Band {
  if (percent === null) return 'unanswered';
  if (percent >= 75) return 'high';
  if (percent >= 50) return 'mid';
  return 'low';
}

export function deriveMaturity(overallPercent: number | null): Maturity {
  if (overallPercent === null) return 'insufficient-data';
  if (overallPercent >= 90) return 'optimising';
  if (overallPercent >= 75) return 'mature';
  if (overallPercent >= 50) return 'developing';
  if (overallPercent >= 25) return 'initial';
  return 'ad-hoc';
}

function scoreSection(
  section: SectionId,
  items: ItemDef[],
  answers: Record<string, string>,
): SectionScore {
  let yesCount = 0;
  let noCount = 0;
  let notApplicableCount = 0;
  let unansweredCount = 0;

  for (const item of items) {
    const answer = answers[item.id] ?? '';
    if (answer === 'yes') yesCount += 1;
    else if (answer === 'no') noCount += 1;
    else if (answer === 'not-applicable') notApplicableCount += 1;
    else unansweredCount += 1;
  }

  const applicableCount = yesCount + noCount;
  const percent = applicableCount === 0 ? null : round2((yesCount / applicableCount) * 100);
  return {
    section,
    yesCount,
    noCount,
    notApplicableCount,
    unansweredCount,
    applicableCount,
    percent,
    band: bandFor(percent),
  };
}

export function calculateMaturity(data: AgileChecklist): GradingResult {
  const teams = scoreSection('teams', TEAMS_ITEMS, data.answers);
  const stakeholders = scoreSection('stakeholders', STAKEHOLDERS_ITEMS, data.answers);
  const practices = scoreSection('practices', PRACTICES_ITEMS, data.answers);

  const answeredCount =
    teams.yesCount + teams.noCount + teams.notApplicableCount +
    stakeholders.yesCount + stakeholders.noCount + stakeholders.notApplicableCount +
    practices.yesCount + practices.noCount + practices.notApplicableCount;

  const sectionPercents = [teams.percent, stakeholders.percent, practices.percent];
  const definedPercents = sectionPercents.filter((p): p is number => p !== null);

  const haveAllSections = definedPercents.length === 3;
  const overallPercent =
    haveAllSections && answeredCount >= MIN_ANSWERED_FOR_REPORT
      ? round2(definedPercents.reduce((a, b) => a + b, 0) / definedPercents.length)
      : null;

  const maturity = deriveMaturity(overallPercent);

  const { firedRules } = applyMaturityRules({ teams, stakeholders, practices });
  const additionalFlags = detectAdditionalFlags(data, {
    teams,
    stakeholders,
    practices,
    answeredCount,
  });

  return {
    answeredCount,
    teams,
    stakeholders,
    practices,
    overallPercent,
    maturity,
    firedRules,
    additionalFlags,
  };
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}
